test/pirates/game_server_test.exs

```diff
@@ -0,0 +1,73 @@
+defmodule Pirates.GameServerTest do
+  use ExUnit.Case, async: true
+  alias Pirates.GameServer
+
+  setup do
+    {:ok, pid} = GameServer.start_link()
+    {:ok, table} = GameServer.register(pid)
+    {:ok, %{pid: pid, table: table}}
+  end
+
+  test "register errors when called again from same process", %{pid: pid} do
+    assert {:error, _} = GameServer.register(pid)
+  end
+
+  test "register always returns the same table for a given server", %{pid: pid, table: original_table} do
+    task = Task.async fn ->
+      {:ok, table} = GameServer.register(pid)
+      table
+    end
+    assert {:ok, ^original_table} = Task.yield(task)
+  end
+
+  test "update_state succeeds when called from a registered process", %{pid: pid, table: table} do
+    assert :ok == GameServer.update_state(table, "foo")
+
+    task = Task.async fn ->
+      {:ok, table} = GameServer.register(pid)
+      GameServer.update_state(table, "foo")
+    end
+    assert {:ok, :ok} == Task.yield(task)
+  end
+
+  test "update_state errors when called from unregistered process", %{table: table} do
+    task = Task.async fn ->
+      GameServer.update_state(table, "foo")
+    end
+    assert {:ok, {:error, _}} = Task.yield(task)
+  end
+
+  test "states returns a list of all states", %{pid: pid, table: table} do
+    assert [] == GameServer.states(table)
+
+    :ok = GameServer.update_state(table, %{id: "foo"})
+    assert [%{id: "foo"}] == GameServer.states(table)
+
+    task = Task.async fn ->
+      {:ok, table} = GameServer.register(pid)
+      GameServer.update_state(table, %{id: "bar"})
+      GameServer.states(table)
+    end
+    results = Task.yield(task)
+    assert {:ok, [%{id: "foo"}, %{id: "bar"}]} == results
+      or   {:ok, [%{id: "bar"}, %{id: "foo"}]} == results # no order to the ets table
+  end
+
+  test "states is automatically 'garbage collected' when a registered process dies", %{pid: pid, table: table} do
+    assert [] == GameServer.states(table)
+
+    :ok = GameServer.update_state(table, %{id: "foo"})
+    assert [%{id: "foo"}] == GameServer.states(table)
+
+    task = Task.async fn ->
+      {:ok, table} = GameServer.register(pid)
+      GameServer.update_state(table, %{id: "bar"})
+    end
+    ref  = Process.monitor(task.pid)
+    assert_receive {:DOWN, ^ref, :process, _, :normal}, 500
+
+    # send a bogus synchronous request to ensure the game server has also processed the :DOWN event
+    GameServer.register(pid)
+    assert [%{id: "foo"}] == GameServer.states(table)
+  end
```

Comment:
Should write test to assert that unregistered processes can access the states info. Probably need to add a function that lets you fetch the table without having to register too...